docs/guillotine.adoc

```diff
@@ -0,0 +1,1349 @@
+
+{zwsp} +
+
+[[guillotine_helpers]]
+=== Guillotine helpers and usage
+
+First off, an introduction to the guillotine helpers at we'll be using. Two of them -  _headless/guillotineApi.es6_ and  _headless/guillotineRequests.es6_ - are general-purpose helpers *included in the react4xp starter*, and the third one we'll write next.
+
+=== Included helper: guillotineApi.es6
+
+The most central of the helpers and the first one we'll use, is *_headless/guillotineApi.es6_*. If we strip away a little boilerplate, the bare essence of it looks like this:
+
+.headless/guillotineApi.es6:
+[source,javascript,options="nowrap"]
+----
+const guillotineLib = require('/lib/guillotine');
+const graphQlLib = require('/lib/graphql');
+
+const SCHEMA = guillotineLib.createSchema();
+
+const executeQuery = (query, variables) =>         <!--1-->
+    graphQlLib.execute(SCHEMA, query, variables);
+
+
+
+// Use in XP controllers:
+exports.executeQuery = executeQuery;                <!--2-->
+
+
+// Expose and use in POST requests from frontend:
+exports.post = req => {                             <!--3-->
+    var body = JSON.parse(req.body);
+
+    return {
+        contentType: 'application/json',
+        body: executeQuery(body.query, body.variables),
+        status: 200
+    };
+};
+----
+
+<1> At the core is the function `executeQuery`. Here, a guillotine `SCHEMA` definition is combined with a graphQL `query` string and an optional `variables` object. These are used with XP's graphQL library to `execute` the query. The result, a JSON object, is returned.
+<2> `executeQuery` is exposed and directly usable from an XP controller. That's what we'll do next.
+<3> a `post` function is also included for receiving POST requests from outside, e.g. a browser. If these requests contain a query string, it's executed with `executeQuery` above, and the result is returned in a response: basically a complete guillotine API endpoint for your webapp.
+
+NOTE: This endpoint is *disabled by default* in the starter, to encourage developers to consider security aspects before using it. We'll get back to that, and activate it, <<#expose_api, later>>.
+
+{zwsp} +
+
+The second included helper, <<#guillotine_request_helper, guillotineRequest.es6>>, is a `fetch` wrapper to simplify guillotine requests at the frontend. We'll take a look at that later.
+
+{zwsp} +
+
+=== Domain-specific helper for listing movies
+
+In order to make requests for a list of movies below a container item in the content hierarchy, we'll need a specific guillotine query string, as well as functionality to adapt the resulting data into the proper props structure for the react components.
+
+And by using the same code on the frontend and backend, for this too, we gain a bit of isomorphism (the predictability of a singe source of truth, in short). So we'll make *a module with custom helper functionality* for our use case, and import from it in both places.
```

Comment:
singe -> single

Reply:
What did you mean?

Reply:
(the predictability of a singe source of truth, in short)

Reply:
Oh, a typo! Hah, thanks!

Singed source of truth... 🔥

Reply:
✅